### PR TITLE
[Dynamo] Support tracing dist.reduce_scatter

### PR DESCRIPTION
**Pull request by guan404ming**
## Related Issue

Part of #151204

## Why

`dist.reduce_scatter` is missing from `traceable_collective_remaps`, so Dynamo inlines its eager implementation and fails on the pybind11 `ReduceScatterOptions()` constructor.

## How

- Remap it to a `reduce_scatter_inplace` wrapper (mirroring `all_gather_inplace`): inputs matching `output`'s shape lower to `funcol.reduce_scatter_single`; mismatched sizes raise, matching eager's own check
- Add tests for the equal-size, scalar, and size-mismatch paths

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta @Lucaskabela

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by guan404ming** [comment]
Added a multi-proc test (`test_reduce_scatter_list_input`, world_size=2) verifying each rank receives the reduction of `input_list[rank]`, compiled with `fullgraph=True` against eager.

**Comment by guilhermeleobas** [comment]
Deferring the review to @aorenste as I am not familiar with distributed code

File: torch/distributed/_functional_collectives.py [hunk 1]
```diff
@@ -1812,6 +1812,30 @@ def all_gather_inplace(
...
+    if not all(t.size() == output.size() for t in input_list):
+        raise AssertionError(
+            "Remapping variable size reduce_scatter is not yet supported"
+        )
+
+    group = group or dist.group.WORLD
+    if group is None:
+        raise AssertionError("group cannot be None")
+
+    return output.copy_(reduce_scatter_single(torch.cat(input_list), op, 0, group, tag))
```
Comment (aorenste) [inline torch/distributed/_functional_collectives.py:1845]: [blocker] Scalar `dist.reduce_scatter` (0-dim inputs and output) passes the size guard above, but `torch.cat` rejects 0-dim tensors ("zero-dimensional tensor ... cannot be concatenated"), and eager accepts such calls -- so this needs a scalar path like `all_gather_inplace`. Stack into 1-D and reshape the per-rank result back (`copy_` won't broadcast `[1]` into a 0-dim output):

```suggestion
    if output.dim() == 0:
        # scalars have no dim to scatter along; stack into 1-D then reshape back
        result = reduce_scatter_single(torch.stack(input_list), op, 0, group, tag)
        return output.copy_(result.reshape(output.shape))
    return output.copy_(reduce_scatter_single(torch.cat(input_list), op, 0, group, tag))
```

File: torch/distributed/_functional_collectives.py [hunk 2]
```diff
@@ -1812,6 +1812,30 @@ def all_gather_inplace(
     return tensor_list
 
 
+def reduce_scatter_inplace(
+    output: torch.Tensor,
+    input_list: list[torch.Tensor],
+    op: str = "sum",
+    group=None,
```
Comment (aorenste) [inline torch/distributed/_functional_collectives.py]: nit:

```suggestion
    group: dist.ProcessGroup | None = None,
```

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[cleanup] Extending the `@claude` bot's still-open note on the uneven path: the "Remapping variable size reduce_scatter is not yet supported" message reads as if reduce_scatter_v were a valid-but-unimplemented case, but eager `dist.reduce_scatter` rejects variable sizes outright -- the backends require every `input_list` element to match `output`'s shape (Gloo: `TORCH_CHECK_EQ(outputs[0].sizes(), inputs[0][i].sizes())`). So `test_dynamo_rewrite_dist_reduce_scatter_list_uneven`'s `[8,4]`/`[4,4]` shapes are not a valid eager call -- they only trip the guard -- and the description's "graph-break to eager otherwise" (fullgraph=False) path is untested and would just hit eager's own size check for such inputs. Consider wording the guard/description as guarding *invalid* input rather than "not yet supported," and either dropping the graph-break claim or covering it with a test.

**Comment by guan404ming** [comment]
Thanks for your review. Reworded the guard as invalid input rather than unimplemented, renamed the test to `_size_mismatch`, and dropped the graph-break claim from the description. Also added a scalar test, since the applied fix had none.

File: torch/distributed/_functional_collectives.py [hunk 3]
```diff
@@ -1997,6 +2034,7 @@ def _remapped_batch_p2p_ops(*args, **kwargs):
     legacy_allreduce: _remapped_allreduce,
     legacy_all_to_all_single: _remapped_all_to_all_single,
     legacy_all_gather: _remapped_all_gather,
+    legacy_reduce_scatter: _remapped_reduce_scatter,
```
Comment (aorenste) [inline torch/distributed/_functional_collectives.py:2037]: [blocker] My agent says: Adding the list-form `reduce_scatter` to `traceable_collective_remaps` routes it through both remap paths, but only the Dynamo path (`functions.py`) was taught to convert `op` from a `ReduceOp` to a lowercase string. `_remap_traceable_collective` (used by non-strict export and `make_fx` under `compile_on_one_rank`) has its own op-conversion tuple that omits the list-form `reduce_scatter`, so an explicitly-passed `op` flows unconverted into `reduce_scatter_single(..., <ReduceOp>, ...)` and crashes at `reduceOp.lower()` with `AttributeError`. Add it to that tuple:

```python
    if func in (
        torch.distributed.all_reduce,
        torch.distributed.reduce_scatter,
        torch.distributed.reduce_scatter_single,
        torch.distributed.reduce_scatter_tensor,
        torch.distributed._reduce_scatter_base,
    ):
```

**Review by aorenste — CHANGES_REQUESTED** [review]
(Reviewed by me, assisted by AI)

[blocker] My agent says: No test exercises `dist.reduce_scatter` through the `make_fx` `compile_on_one_rank` / non-strict-export remap path or with a non-default `op` -- every current test uses the default op and only the `torch.compile` (Dynamo) path. Mirror `TestCompileOnOneRankLegacyCollective.test_legacy_all_reduce_serializes_under_coor` in `test/distributed/tensor/test_compile_on_one_rank.py` for `reduce_scatter` with `op=dist.ReduceOp.MAX` (fake PG, no GPU needed); that test fails today at `reduceOp.lower()` and passes once `reduce_scatter` is added to the op-conversion tuple.

**Comment by guan404ming** [comment]
Thanks for you review. Both fixed: added `reduce_scatter` to `_remap_traceable_collective`'s op-conversion tuple, and added the COOR test with op=MAX (verified fail-before/pass-after). PTAL.

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

**Comment by guan404ming** [comment]
Thanks!
